File: typescript/src/doorExample.ts
// Useful functions
function logDoorState(door: Door): void {
  if (door.state === "open") {
    console.log("door is opened");
    return;
  }

  if (door.isLocked) {
    console.log("door is closed and locked");
    return;
  }

  console.log("door is locked but unlocked");
}

// Types
interface OpenDoor {
  state: "open";
}

interface CloseDoor {
  state: "closed",
  isLocked: boolean
}

type Door = OpenDoor | CloseDoor

// Usage
const door1: Door = { state: "open" };
const door2: Door = { state: "closed", isLocked: false };
const door3: Door = { state: "closed", isLocked: true };

logDoorState(door1);
logDoorState(door2);
logDoorState(door3);

// These won't be accepted by tsc
// const door4: Door = {state: "open", isLocked: false};
// const door5: Door = {state: "open", isLocked: true};
// const door6: Door = {state: "closed"};

File: python/src/examples/door_example.py
from dataclasses import dataclass
from typing import Literal, Protocol


# Types
@dataclass
class OpenDoor:
    pass


@dataclass
class ClosedDoor:
    is_locked: bool


type Door = OpenDoor | ClosedDoor


def log_door_state(door: Door) -> None:
    if isinstance(door, OpenDoor):
        print("Door opened")
        return

    print(f"Door closed {'and locked' if door.is_locked else 'unlocked'}")


if __name__ == "__main__":
    # Usage
    door1: Door = OpenDoor()
    door2: Door = ClosedDoor(is_locked=False)
    door3: Door = ClosedDoor(is_locked=True)

    log_door_state(door1)
    log_door_state(door2)
    log_door_state(door3)
